CPMemberAddRemoveTest.when_newCPMemberIsAddedToTheMetadataGroupAfterRestartAndSnapshot_newMemberInstallsSnapshot

http://jenkins.hazelcast.com/job/Hazelcast-pr-builder/3808

```
java.lang.IllegalStateException: Missing CP member in active CP members: [CPMember{uuid=cd80dc0e-ea1a-466f-8bf5-620212d08bb0, address=[127.0.0.1]:5704}, CPMember{uuid=f4c54513-6aaf-41c2-9173-0140f54b7f5e, address=[127.0.0.1]:5705}, CPMember{uuid=539027f8-1688-4270-bdd3-3c5f4e552e0e, address=[127.0.0.1]:5701}] for CPGroupInfo{id=CPGroupId{name='METADATA', seed=1571122220030, commitIndex=0}, initialMembers=[RaftEndpoint{uuid='cd80dc0e-ea1a-466f-8bf5-620212d08bb0'}, RaftEndpoint{uuid='f4c54513-6aaf-41c2-9173-0140f54b7f5e'}, RaftEndpoint{uuid='539027f8-1688-4270-bdd3-3c5f4e552e0e'}], membersCommitIndex=65, members=[RaftEndpoint{uuid='cd80dc0e-ea1a-466f-8bf5-620212d08bb0'}, RaftEndpoint{uuid='f4c54513-6aaf-41c2-9173-0140f54b7f5e'}, RaftEndpoint{uuid='7665b5e7-f5e3-4bc7-87a8-2e626ee6c5c9'}], status=ACTIVE}
	at com.hazelcast.cp.internal.CPGroupInfo.toSummary(CPGroupInfo.java:169)
	at com.hazelcast.cp.internal.MetadataRaftGroupManager.getGroup(MetadataRaftGroupManager.java:344)
	at com.hazelcast.cp.internal.HazelcastRaftTestSupport.getRaftGroupLocally(HazelcastRaftTestSupport.java:220)
	at com.hazelcast.cp.internal.CPMemberAddRemoveTest.lambda$when_newCPMemberIsAddedToTheMetadataGroupAfterRestartAndSnapshot_newMemberInstallsSnapshot$19(CPMemberAddRemoveTest.java:815)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1316)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1418)
	at com.hazelcast.cp.internal.CPMemberAddRemoveTest.when_newCPMemberIsAddedToTheMetadataGroupAfterRestartAndSnapshot_newMemberInstallsSnapshot(CPMemberAddRemoveTest.java:811)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:114)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```

```
java.util.concurrent.ExecutionException: java.lang.IllegalStateException: Cannot rebalance CP groups because there is ongoing MembershipChangeSchedule{membershipChangeCommitIndices=[58], member=CPMember{uuid=e270e361-6789-4c12-8e25-ec0f8752ff12, address=[127.0.0.1]:5701}, membershipChangeMode=REMOVE, changes=[CPGroupMembershipChange{groupId=CPGroupId{name='group1', seed=1572504503473, commitIndex=56}, membersCommitIndex=0, members=[RaftEndpoint{uuid='cf2f46d7-531f-4db0-abf2-542bdd36cae3'}, RaftEndpoint{uuid='7a8c77a0-e3dc-4828-a931-7aacea5d04cf'}, RaftEndpoint{uuid='e270e361-6789-4c12-8e25-ec0f8752ff12'}], memberToAdd=null, memberToRemove=RaftEndpoint{uuid='e270e361-6789-4c12-8e25-ec0f8752ff12'}}, CPGroupMembershipChange{groupId=CPGroupId{name='METADATA', seed=1572504503473, commitIndex=0}, membersCommitIndex=0, members=[RaftEndpoint{uuid='cf2f46d7-531f-4db0-abf2-542bdd36cae3'}, RaftEndpoint{uuid='e270e361-6789-4c12-8e25-ec0f8752ff12'}, RaftEndpoint{uuid='7a8c77a0-e3dc-4828-a931-7aacea5d04cf'}], memberToAdd=null, memberToRemove=RaftEndpoint{uuid='e270e361-6789-4c12-8e25-ec0f8752ff12'}}, CPGroupMembershipChange{groupId=CPGroupId{name='group2', seed=1572504503473, commitIndex=57}, membersCommitIndex=0, members=[RaftEndpoint{uuid='e270e361-6789-4c12-8e25-ec0f8752ff12'}, RaftEndpoint{uuid='7a8c77a0-e3dc-4828-a931-7aacea5d04cf'}, RaftEndpoint{uuid='cf2f46d7-531f-4db0-abf2-542bdd36cae3'}], memberToAdd=null, memberToRemove=RaftEndpoint{uuid='e270e361-6789-4c12-8e25-ec0f8752ff12'}}]}
	at java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:357)
	at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1895)
	at com.hazelcast.spi.impl.DeserializingCompletableFuture.get(DeserializingCompletableFuture.java:96)
	at com.hazelcast.cp.internal.CPMemberAddRemoveTest.when_newCPMemberIsAddedToTheMetadataGroupAfterRestartAndSnapshot_newMemberInstallsSnapshot(CPMemberAddRemoveTest.java:877)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:114)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.IllegalStateException: Cannot rebalance CP groups because there is ongoing MembershipChangeSchedule{membershipChangeCommitIndices=[58], member=CPMember{uuid=e270e361-6789-4c12-8e25-ec0f8752ff12, address=[127.0.0.1]:5701}, membershipChangeMode=REMOVE, changes=[CPGroupMembershipChange{groupId=CPGroupId{name='group1', seed=1572504503473, commitIndex=56}, membersCommitIndex=0, members=[RaftEndpoint{uuid='cf2f46d7-531f-4db0-abf2-542bdd36cae3'}, RaftEndpoint{uuid='7a8c77a0-e3dc-4828-a931-7aacea5d04cf'}, RaftEndpoint{uuid='e270e361-6789-4c12-8e25-ec0f8752ff12'}], memberToAdd=null, memberToRemove=RaftEndpoint{uuid='e270e361-6789-4c12-8e25-ec0f8752ff12'}}, CPGroupMembershipChange{groupId=CPGroupId{name='METADATA', seed=1572504503473, commitIndex=0}, membersCommitIndex=0, members=[RaftEndpoint{uuid='cf2f46d7-531f-4db0-abf2-542bdd36cae3'}, RaftEndpoint{uuid='e270e361-6789-4c12-8e25-ec0f8752ff12'}, RaftEndpoint{uuid='7a8c77a0-e3dc-4828-a931-7aacea5d04cf'}], memberToAdd=null, memberToRemove=RaftEndpoint{uuid='e270e361-6789-4c12-8e25-ec0f8752ff12'}}, CPGroupMembershipChange{groupId=CPGroupId{name='group2', seed=1572504503473, commitIndex=57}, membersCommitIndex=0, members=[RaftEndpoint{uuid='e270e361-6789-4c12-8e25-ec0f8752ff12'}, RaftEndpoint{uuid='7a8c77a0-e3dc-4828-a931-7aacea5d04cf'}, RaftEndpoint{uuid='cf2f46d7-531f-4db0-abf2-542bdd36cae3'}], memberToAdd=null, memberToRemove=RaftEndpoint{uuid='e270e361-6789-4c12-8e25-ec0f8752ff12'}}]}
	at com.hazelcast.internal.util.Preconditions.checkState(Preconditions.java:355)
	at com.hazelcast.cp.internal.MetadataRaftGroupManager.addMember(MetadataRaftGroupManager.java:1010)
	at com.hazelcast.cp.internal.raftop.metadata.AddCPMemberOp.run(AddCPMemberOp.java:50)
	at com.hazelcast.cp.internal.raftop.metadata.MetadataRaftGroupOp.run(MetadataRaftGroupOp.java:38)
	at com.hazelcast.cp.internal.NodeEngineRaftIntegration.runOperation(NodeEngineRaftIntegration.java:210)
	at com.hazelcast.cp.internal.raft.impl.RaftNodeImpl.applyLogEntry(RaftNodeImpl.java:828)
	at com.hazelcast.cp.internal.raft.impl.RaftNodeImpl.applyLogEntries(RaftNodeImpl.java:769)
	at com.hazelcast.cp.internal.raft.impl.RaftNodeImpl.commitEntries(RaftNodeImpl.java:1278)
	at com.hazelcast.cp.internal.raft.impl.RaftNodeImpl.tryAdvanceCommitIndex(RaftNodeImpl.java:1262)
	at com.hazelcast.cp.internal.raft.impl.handler.AppendSuccessResponseHandlerTask.handleResponse(AppendSuccessResponseHandlerTask.java:71)
	at com.hazelcast.cp.internal.raft.impl.handler.AbstractResponseHandlerTask.innerRun(AbstractResponseHandlerTask.java:43)
	at com.hazelcast.cp.internal.raft.impl.task.RaftNodeStatusAwareTask.run(RaftNodeStatusAwareTask.java:52)
	at com.hazelcast.cp.internal.NodeEngineRaftIntegration.execute(NodeEngineRaftIntegration.java:113)
	at com.hazelcast.cp.internal.raft.impl.RaftNodeImpl.execute(RaftNodeImpl.java:862)
	at com.hazelcast.cp.internal.raft.impl.RaftNodeImpl.handleAppendResponse(RaftNodeImpl.java:393)
	at com.hazelcast.cp.internal.RaftService.handleAppendResponse(RaftService.java:645)
	at com.hazelcast.cp.internal.operation.integration.AppendSuccessResponseOp.run(AppendSuccessResponseOp.java:48)
	at com.hazelcast.spi.impl.operationservice.Operation.call(Operation.java:184)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.call(OperationRunnerImpl.java:216)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:205)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:429)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:154)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:124)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.run(OperationThread.java:111)
```

http://jenkins.hazelcast.com/job/Hazelcast-pr-builder/4066/testReport/junit/com.hazelcast.cp.internal/CPMemberAddRemoveTest/when_newCPMemberIsAddedToTheMetadataGroupAfterRestartAndSnapshot_newMemberInstallsSnapshot/

--- a/hazelcast/src/main/java/com/hazelcast/cp/internal/RaftService.java
+++ b/hazelcast/src/main/java/com/hazelcast/cp/internal/RaftService.java
@@ -142,7 +142,7 @@ public class RaftService implements ManagedService, SnapshotAwareService<Metadat
 
     public static final String SERVICE_NAME = "hz:core:raft";
 
-    static final String CP_SUBSYSTEM_EXECUTOR = "hz:cpSubsystem";
+    public static final String CP_SUBSYSTEM_EXECUTOR = "hz:cpSubsystem";
     static final String CP_SUBSYSTEM_MANAGEMENT_EXECUTOR = "hz:cpSubsystemManagement";
 
     private static final long REMOVE_MISSING_MEMBER_TASK_PERIOD_SECONDS = 1;

--- a/hazelcast/src/main/java/com/hazelcast/cp/internal/operation/RaftQueryOp.java
+++ b/hazelcast/src/main/java/com/hazelcast/cp/internal/operation/RaftQueryOp.java
@@ -36,6 +36,7 @@ import com.hazelcast.spi.impl.operationservice.Operation;
 import java.io.IOException;
 import java.util.function.BiConsumer;
 
+import static com.hazelcast.cp.internal.RaftService.CP_SUBSYSTEM_EXECUTOR;
 import static com.hazelcast.internal.util.InvocationUtil.CALLER_RUNS_EXECUTOR;
 
 /**
@@ -76,8 +77,8 @@ public class RaftQueryOp extends Operation implements IndeterminateOperationStat
             }
             return;
         } else if (raftNode.getStatus() == RaftNodeStatus.STEPPED_DOWN) {
-            service.stepDownRaftNode(groupId);
             sendResponse(new NotLeaderException(groupId, service.getLocalCPEndpoint(), null));
+            getNodeEngine().getExecutionService().execute(CP_SUBSYSTEM_EXECUTOR, () -> service.stepDownRaftNode(groupId));
             return;
         }
 

--- a/hazelcast/src/main/java/com/hazelcast/cp/internal/operation/RaftReplicateOp.java
+++ b/hazelcast/src/main/java/com/hazelcast/cp/internal/operation/RaftReplicateOp.java
@@ -34,7 +34,7 @@ import java.io.IOException;
 import java.util.concurrent.Executor;
 import java.util.function.BiConsumer;
 
-import static com.hazelcast.internal.util.InvocationUtil.CALLER_RUNS_EXECUTOR;
+import static com.hazelcast.cp.internal.RaftService.CP_SUBSYSTEM_EXECUTOR;
 
 /**
  * The base class that replicates the given {@link RaftOp}
@@ -70,8 +70,8 @@ public abstract class RaftReplicateOp extends Operation implements IdentifiedDat
             }
             return;
         } else if (raftNode.getStatus() == RaftNodeStatus.STEPPED_DOWN) {
-            service.stepDownRaftNode(groupId);
             sendResponse(new NotLeaderException(groupId, service.getLocalCPEndpoint(), null));
+            getNodeEngine().getExecutionService().execute(CP_SUBSYSTEM_EXECUTOR, () -> service.stepDownRaftNode(groupId));
             return;
         }
 

--- a/hazelcast/src/main/java/com/hazelcast/cp/internal/raftop/metadata/CompleteDestroyRaftGroupsOp.java
+++ b/hazelcast/src/main/java/com/hazelcast/cp/internal/raftop/metadata/CompleteDestroyRaftGroupsOp.java
@@ -77,7 +77,7 @@ public class CompleteDestroyRaftGroupsOp extends MetadataRaftGroupOp implements 
     @Override
     public void readData(ObjectDataInput in) throws IOException {
         int count = in.readInt();
-        groupIds = new HashSet<CPGroupId>();
+        groupIds = new HashSet<>();
         for (int i = 0; i < count; i++) {
             CPGroupId groupId = in.readObject();
             groupIds.add(groupId);

--- a/hazelcast/src/main/java/com/hazelcast/cp/internal/raftop/metadata/CompleteRaftGroupMembershipChangesOp.java
+++ b/hazelcast/src/main/java/com/hazelcast/cp/internal/raftop/metadata/CompleteRaftGroupMembershipChangesOp.java
@@ -81,7 +81,7 @@ public class CompleteRaftGroupMembershipChangesOp extends MetadataRaftGroupOp im
     @Override
     public void readData(ObjectDataInput in) throws IOException {
         int count = in.readInt();
-        changedGroups = new HashMap<CPGroupId, BiTuple<Long, Long>>(count);
+        changedGroups = new HashMap<>(count);
         for (int i = 0; i < count; i++) {
             CPGroupId groupId = in.readObject();
             long currMembersCommitIndex = in.readLong();

--- a/hazelcast/src/main/java/com/hazelcast/cp/internal/raftop/metadata/CreateRaftGroupOp.java
+++ b/hazelcast/src/main/java/com/hazelcast/cp/internal/raftop/metadata/CreateRaftGroupOp.java
@@ -93,7 +93,7 @@ public class CreateRaftGroupOp extends MetadataRaftGroupOp implements Indetermin
     public void readData(ObjectDataInput in) throws IOException {
         groupName = in.readUTF();
         int len = in.readInt();
-        members = new ArrayList<RaftEndpoint>(len);
+        members = new ArrayList<>(len);
         for (int i = 0; i < len; i++) {
             RaftEndpoint member = in.readObject();
             members.add(member);

--- a/hazelcast/src/main/java/com/hazelcast/cp/internal/raftop/metadata/CreateRaftNodeOp.java
+++ b/hazelcast/src/main/java/com/hazelcast/cp/internal/raftop/metadata/CreateRaftNodeOp.java
@@ -95,7 +95,7 @@ public class CreateRaftNodeOp extends Operation implements IdentifiedDataSeriali
         super.readInternal(in);
         groupId = in.readObject();
         int count = in.readInt();
-        initialMembers = new ArrayList<RaftEndpoint>(count);
+        initialMembers = new ArrayList<>(count);
         for (int i = 0; i < count; i++) {
             RaftEndpoint member = in.readObject();
             initialMembers.add(member);

--- a/hazelcast/src/main/java/com/hazelcast/cp/internal/raftop/metadata/InitMetadataRaftGroupOp.java
+++ b/hazelcast/src/main/java/com/hazelcast/cp/internal/raftop/metadata/InitMetadataRaftGroupOp.java
@@ -89,7 +89,7 @@ public class InitMetadataRaftGroupOp extends MetadataRaftGroupOp implements Inde
     public void readData(ObjectDataInput in) throws IOException {
         callerCPMember = in.readObject();
         int len = in.readInt();
-        discoveredCPMembers = new ArrayList<CPMemberInfo>(len);
+        discoveredCPMembers = new ArrayList<>(len);
         for (int i = 0; i < len; i++) {
             CPMemberInfo member = in.readObject();
             discoveredCPMembers.add(member);

--- a/hazelcast/src/main/java/com/hazelcast/cp/internal/raftop/metadata/PublishActiveCPMembersOp.java
+++ b/hazelcast/src/main/java/com/hazelcast/cp/internal/raftop/metadata/PublishActiveCPMembersOp.java
@@ -96,7 +96,7 @@ public class PublishActiveCPMembersOp extends Operation implements IdentifiedDat
         metadataGroupId = in.readObject();
         membersCommitIndex = in.readLong();
         int len = in.readInt();
-        members = new ArrayList<CPMemberInfo>(len);
+        members = new ArrayList<>(len);
         for (int i = 0; i < len; i++) {
             CPMemberInfo member = in.readObject();
             members.add(member);

--- a/hazelcast/src/main/java/com/hazelcast/cp/internal/raftop/metadata/RemoveCPMemberOp.java
+++ b/hazelcast/src/main/java/com/hazelcast/cp/internal/raftop/metadata/RemoveCPMemberOp.java
@@ -27,10 +27,11 @@ import com.hazelcast.nio.ObjectDataOutput;
 import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
 
 import java.io.IOException;
+import java.util.UUID;
 
 /**
  * When a CP member is shutting down gracefully, or a crashed CP member is
- * removed from CP Subsystem via {@link RaftService#removeCPMember(String)},
+ * removed from CP Subsystem via {@link RaftService#removeCPMember(UUID)},
  * this operation is committed to the Metadata Raft group.
  */
 public class RemoveCPMemberOp extends MetadataRaftGroupOp implements IndeterminateOperationStateAware,

--- a/hazelcast/src/main/java/com/hazelcast/cp/internal/raftop/metadata/TerminateRaftNodesOp.java
+++ b/hazelcast/src/main/java/com/hazelcast/cp/internal/raftop/metadata/TerminateRaftNodesOp.java
@@ -91,7 +91,7 @@ public class TerminateRaftNodesOp extends Operation implements IdentifiedDataSer
     protected void readInternal(ObjectDataInput in) throws IOException {
         super.readInternal(in);
         int count = in.readInt();
-        groupIds = new ArrayList<CPGroupId>();
+        groupIds = new ArrayList<>();
         for (int i = 0; i < count; i++) {
             CPGroupId groupId = in.readObject();
             groupIds.add(groupId);

